doc/adr/0007-marconi-indexer.rst

```diff
@@ -0,0 +1,118 @@
+
+The slot number information is used in case of rollbacks (when we only get the old slot number that we need to rollback to) and for resuming the operation of the index, in which case we need both the slot numbers and block ids.
+
+Note that to support resume from disk we need to always have at least one event persisted on disk which contains the slot number and block id from which we are supposed to resume operation. In case there are more than one events stored on disk we can use those as resume alternatives in the case of the ChainSync protocol.
+
+Queries
+^^^^^^^
+
+We also extend the queries with a bit more structure that will make specifying query validity intervals possible. The validity is important in the case where we want to query several indexers and we would like them to require to have processed all the information up to some slot number or be between some slot interval.
+
+At this point all query results are synchronous. We have plans to extend this functionality, but these plans are based on updating the notification system for indexers which will be described in a further ADR.
+
+API Design
+^^^^^^^^^^
+
+We want the API for our users to be as flexible as possible, so some of our previously mentioned design patterns are not captured by the API, but rather by its implementation.
+
+Data types
+""""""""""
+
+* Events
+  ** Slot numbers (requires Ord)
+  ** Block id (requires Eq)
+  ** e (type variable standing for the event)
+* Query
+  ** Validity interval (can be always valid)
+  ** q (type variable standing for the query)
+* Result
+  ** Slot number at which the query was ran
+  ** r (type variable standing for the query result)
+
+Functions
+"""""""""
+
+* Query
+  ** Indexer
+  ** Validity interval
+  ** The query (q type variable)
+  ** Returns the result
+* Store
+  ** Indexer
+  ** Does not return anything useful
+* Resume
+  ** Indexer
+  ** Returns a list of slot numbers and block ids
+
+Runtime parameters
+""""""""""""""""""
+
+* Minimum events retained (this should be the previously mentioned K parameter)
+* Maximum in-memory events (should be less than K)
+
+Extension mechanisms
+""""""""""""""""""""
+
+A. Storage engine
+
+You can customise the query and store functions which run in some generic monad to use whatever backend is best for the job. We currently use SQLite, but that is more for convenience than anything else.
+
+B. Query intervals
+
+If you want to specify an interval for your queries (which is highly encouraged) then you need to have either in memory or on disk sufficient information to reconstruct the state at the given slot number. By storing more than K events you can extend the query interval as much as you need. In extreme, you can only store events on disk, in which case your queries can span the whole blockchain.
+
+Implementation
+^^^^^^^^^^^^^^
+
+This is the way we suggest people implement storage for the indexers.::
```

Comment:
indexers.:: -> indexers:

Reply:
I think it is correct though. I would like the following paragraph to be taken verbatim.

Reply:
It's just that the rendering show "indexers.:". There's an extra dot.

Reply:
The way that `::` is handled is apparently 'smart' (unpredictable would be the word I would use) and it should be fixed now.